### PR TITLE
Add search to the cards page!

### PR DESCRIPTION
**Pull request by sampoder**
## Summary of the problem

Closes https://github.com/hackclub/hcb/issues/4975

## Describe your changes

https://github.com/hackclub/hcb/assets/39828164/cc363d9c-ef40-4ee0-918f-7f1dfa789a73

Is it a security issue to have search by last four? Should I disable that?

### CODE REVIEW

**Comment by garyhtou** [comment]
Ideally, you should be able to search by last4 for only your cards. But that's a bit complicated to put together. So maybe we remove that feature for now

File: app/controllers/events_controller.rb [hunk 1]
```diff
@@ -270,10 +270,12 @@ def emburse_card_overview
   end
 
   def card_overview
+    @q = params[:q].presence
     @stripe_cards = @event.stripe_cards.includes(:stripe_cardholder, :user).order("created_at desc")
+    @stripe_cards = @stripe_cards.joins(stripe_cardholder: :user).where("users.full_name ILIKE :query OR users.email ILIKE :query OR stripe_cards.last4 ILIKE :query", query: "%#{User.sanitize_sql_like(@q)}%") if @q
```
Comment (sampoder) [inline app/controllers/events_controller.rb]:
```suggestion
    @stripe_cards = @stripe_cards.joins(stripe_cardholder: :user).where("users.full_name ILIKE :query OR users.email ILIKE :query, query: "%#{User.sanitize_sql_like(@q)}%") if @q
```
let's remove it then yeah

File: app/views/events/card_overview.html.erb [hunk 2]
```diff
@@ -79,7 +79,13 @@
     </section>
   <% end %>
 
-  <h4>Organization Cards</h4>
+  <% unless @q %>
+    <h3 class="mb2">Organization Cards</h3>
+  <% end %>
+
+  <%= form_with method: :get do |form| %>
+    <%= form.text_field :q, value: @q, placeholder: "Search by cardholder / card numbers", class: "mb3 w-100", style: "max-width: 100%" %>
```
Comment (sampoder) [inline app/views/events/card_overview.html.erb]:
```suggestion
    <%= form.text_field :q, value: @q, placeholder: "Search by cardholder", class: "mb3 w-100", style: "max-width: 100%" %>
```

File: app/controllers/events_controller.rb [hunk 3]
```diff
@@ -270,10 +270,12 @@ def emburse_card_overview
   end
 
   def card_overview
+    @q = params[:q].presence
     @stripe_cards = @event.stripe_cards.includes(:stripe_cardholder, :user).order("created_at desc")
+    @stripe_cards = @stripe_cards.joins(stripe_cardholder: :user).where("users.full_name ILIKE :query OR users.email ILIKE :query, query: "%#{User.sanitize_sql_like(@q)}%") if @q
```
Comment (garyhtou) [inline app/controllers/events_controller.rb]: i'd be nice to search by card name

https://github.com/hackclub/bank/blob/dd9bdb4b8754a0ac276b688082d16b7f145d5065/app/models/stripe_card.rb#L13

Comment (sampoder) [inline app/controllers/events_controller.rb]:
```suggestion
    @stripe_cards = @stripe_cards.joins(stripe_cardholder: :user).where("users.full_name ILIKE :query OR users.email ILIKE :query OR stripe_cards.name ILIKE :query, query: "%#{User.sanitize_sql_like(@q)}%") if @q
```

**Pull request merged**